core/sql/generator/GenPreCode.cpp

```diff
@@ -11274,6 +11274,485 @@ short HbaseAccess::extractHbaseFilterPreds(Generator * generator,
+			  op = "LESS"+nullType;
+		  else
+			  op = "GREATER"+nullType;
+	  }else  if (ie->getOperatorType() == ITM_GREATER_EQ){
+		  if (flipOp)
+			  op = "LESS_OR_EQUAL"+nullType;
+		  else
+			  op = "GREATER_OR_EQUAL"+nullType;
+	  }else
+	    op = "NO_OP"+nullType;
+	}
+    }
+  if (foundUnary){
+	  const NAType &colType = colVID.getType();
+	  NAString nullType;
+
+	  if (colType.supportsSQLnull())
+	    {
+	      nullType = "_NULL";
+	    }
+	  else
+	    {
+	      nullType = "";
+	    }
+	  if (ie->getOperatorType() == ITM_IS_NULL)
+	  		  op = "IS_NULL"+nullType;
+	  else if (ie->getOperatorType() == ITM_IS_NOT_NULL)
+		      op = "IS_NOT_NULL"+nullType;
+  }
+
+  return foundBinary || foundUnary;
+}
+short HbaseAccess::extractHbaseFilterPredsVX(Generator * generator,
+		   ValueIdSet &preds, ValueIdSet &newExePreds){
+	//separate the code that should not belong in the recursive function
+	   if (CmpCommon::getDefault(HBASE_FILTER_PREDS) == DF_OFF)
+	    return 0;
+	   // check if initial (version 1) implementation
+	   if (CmpCommon::getDefault(HBASE_FILTER_PREDS) == DF_MINIMUM)
+	    return extractHbaseFilterPreds(generator,preds,newExePreds);
+
+	   // if here, we are DF_MEDIUM
+	   // cannot push preds for aligned format row
+	   NABoolean isAlignedFormat = getTableDesc()->getNATable()->isAlignedFormat(getIndexDesc());
+
+	   if (isAlignedFormat)
+	     return 0;
+	   //recursive function call
+	   opList_.insert("V2");//to instruct the java side that we are dealing with predicate pushdown V2 semantic, add "V2" marker
+	   extractHbaseFilterPredsV2(generator,preds,newExePreds,FALSE,TRUE);
+	   return 0;
+
+}
+
+// return true if successfull push down of node
+NABoolean HbaseAccess::extractHbaseFilterPredsV2(Generator * generator,
+					   ValueIdSet &preds, ValueIdSet &newExePreds, NABoolean checkOnly, NABoolean isFirstAndLayer )
+{
+
+	// the isFirstAndLayer is used to allow detecting top level predicate that can still be pushed to executor
+	int addedNode=0;
+	for (ValueId vid = preds.init();
+       (preds.next(vid));
+       preds.advance(vid))
+    {
+      ItemExpr * ie = vid.getItemExpr();
+
+      // if it is AND operation, recurse through left and right children
+      if (ie->getOperatorType() == ITM_AND)
+        {
+          ValueIdSet leftPreds;
+          ValueIdSet rightPreds;
+          leftPreds += ie->child(0)->castToItemExpr()->getValueId();
+          rightPreds += ie->child(1)->castToItemExpr()->getValueId();
+          if (isFirstAndLayer){
+            NABoolean leftOK  = extractHbaseFilterPredsV2(generator, leftPreds, newExePreds,TRUE, TRUE);
+            NABoolean rightOK = extractHbaseFilterPredsV2(generator, rightPreds, newExePreds,TRUE, TRUE);
+            if (leftOK && rightOK){
+            	if (!checkOnly){
+					extractHbaseFilterPredsV2(generator, leftPreds, newExePreds,FALSE, TRUE);//generate tree
+					extractHbaseFilterPredsV2(generator, rightPreds, newExePreds,FALSE, TRUE);//generate tree
```

Comment:
Which is the reason for a split. Sorry.

On Tue, Jan 12, 2016 at 10:44 AM, Eric Owhadi <notifications@github.com>
wrote:

> In core/sql/generator/GenPreCode.cpp
> <https://github.com/apache/incubator-trafodion/pull/255#discussion_r49479048>
> :
>
> > +      ItemExpr * ie = vid.getItemExpr();
> > +
> > +      // if it is AND operation, recurse through left and right children
> > +      if (ie->getOperatorType() == ITM_AND)
> > +        {
> > +          ValueIdSet leftPreds;
> > +          ValueIdSet rightPreds;
> > +          leftPreds += ie->child(0)->castToItemExpr()->getValueId();
> > +          rightPreds += ie->child(1)->castToItemExpr()->getValueId();
> > +          if (isFirstAndLayer){
> > +            NABoolean leftOK  = extractHbaseFilterPredsV2(generator, leftPreds, newExePreds,TRUE, TRUE);
> > +            NABoolean rightOK = extractHbaseFilterPredsV2(generator, rightPreds, newExePreds,TRUE, TRUE);
> > +            if (leftOK && rightOK){
> > +            	if (!checkOnly){
> > +					extractHbaseFilterPredsV2(generator, leftPreds, newExePreds,FALSE, TRUE);//generate tree
> > +					extractHbaseFilterPredsV2(generator, rightPreds, newExePreds,FALSE, TRUE);//generate tree
>
> this function is way too big to duplicate. it will be error prone to split
> in 2.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-trafodion/pull/255/files#r49479048>.
>



-- 
Regards, --Qifan